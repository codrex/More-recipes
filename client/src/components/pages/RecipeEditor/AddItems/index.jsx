import React from 'react';
import PropTypes from 'prop-types';
import ItemsList from './ItemsList';
import { Input, Textarea, Form } from '../../../common/FormElements';
import { validateItem } from '../../../../utils/validator';
import ItemField from './ItemField';
import { RECIPE_ADDED } from '../../../../constants';

/**
 * React component to add recipe Items
 */
class AddItems extends React.Component {
  /**
   * constructor
   * @param {object} props
   */
  constructor(props) {
    super(props);
    this.state = {
      items: props.items || [],
      showInput: true,
      validationError: props.externalError
    };
  }

  /**
   * @param {Object} nextProps
   * @return {boolean} boolean value
   */
  componentWillReceiveProps = (nextProps) => {
    if (this.state.validationError !== nextProps.externalError) {
      this.setState({
        validationError: nextProps.externalError
      });
    }
    this.setState({
      items: nextProps.items
    });
  }

  /**
   * @param {object} nextProps
   * @return {boolean} boolean value
   */
  shouldComponentUpdate = (nextProps) => {
    if (nextProps.message === RECIPE_ADDED) {
      return false;
    }
    return true;
  }

  /**
   * add item to list
   * @param {String} value
   * @return {undefined} undefined
   */
  addItem = (value) => {
    const { directions, ingredients } = this.props;
    const { items } = this.state;
    const item = value[this.props.name].toLowerCase();

    const itemsUpdate = (ingredients && [item].concat(items)) ||
    (directions && items.concat(item));

    this.props.sendItemsToStore(itemsUpdate);
    this.props.initialize();
  }

  /**
   * remove item from list
   * @param {number} index
   * @return {undefined} undefined
   */
  deleteFromList = (index) => {
    const items = [].concat(this.state.items);
    items.splice(index, 1);
    this.props.sendItemsToStore(items);
  }

  /**
   * edit item
   * @param {string} value
   * @param {number} index
   * @return {undefined} undefined
   */
  editItems = (value, index) => {
    const items = [].concat(this.state.items);
    items[index] = value;
    this.props.sendItemsToStore(items);
  }

  /**
   * validate item
   * @return {Array} array of strings
   * @param {string} value
   */
  validate = (value) => {
    const { name } = this.props;
    const lowerCaseValue = value && value.toLowerCase();
    if (this.state.items.includes(lowerCaseValue)) {
      return [`${name} is already on the list`];
    }
    return validateItem(lowerCaseValue, name);
  }

  /**
   * render form
   * @return {React} react component
   */
  renderForm = () => {
    const {
      handleSubmit,
      name,
      placeholder,
      ingredients,
      directions,
      clearValidationError
    } = this.props;
    const Component = (ingredients && Input) || (directions && Textarea);
    if (this.state.showInput) {
      return (
        <Form
          onSubmit={handleSubmit(this.addItem)}
          submitBtnText={`add ${name}`}
          secondary
          lg={false}
          id={`${name}-submit-btn`}
        >
          <ItemField
            Component={Component}
            name={name}
            placeholder={placeholder}
            validate={this.validate}
            clearValidationError={clearValidationError}
            validationError={this.state.validationError}
          />
        </Form>
      );
    } return null;
  }

  /**
   * render list
   * @return {React} react component
   */
  renderList = () => {
    const {
      name,
      Component,
      ingredients,
      directions
    } = this.props;
    return (
      <ItemsList
        items={this.state.items}
        deleteItem={this.deleteFromList}
        editItem={this.editItems}
        componentType={name}
        Component={Component}
        name={name}
        directions={directions}
        ingredients={ingredients}
        uniquenessChecker={this.validate}
      />
    );
  }

  /**
   * render
   * @return {React} react component
   */
  render = () => {
    const { name } = this.props;
    const { items } = this.state;
    return (
      <div className="col-12 items ">
        <div className="items-header">
          <h4 className="items-header-text">
            {`${name}s ${' '}${items.length}`}
          </h4>
          {this.renderForm()}
        </div>
        <div className="col-12 items-list-wrapper">
          {this.renderList()}
        </div>
      </div>
    );
  }
}

AddItems.defaultProps = {
  ingredients: false,
  directions: false,
  externalError: [],
  Component: null
};

AddItems.propTypes = {
  clearValidationError: PropTypes.func.isRequired,
  directions: PropTypes.bool,
  externalError: PropTypes.oneOfType(
    [
      PropTypes.string,
      PropTypes.array, PropTypes.bool
    ]
  ),
  handleSubmit: PropTypes.func.isRequired,
  ingredients: PropTypes.bool,
  initialize: PropTypes.func.isRequired,
  items: PropTypes.arrayOf(PropTypes.shape).isRequired,
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string.isRequired,
  sendItemsToStore: PropTypes.func.isRequired,
  Component: PropTypes.oneOfType(
    [
      PropTypes.node,
      PropTypes.element,
      PropTypes.func
    ]
  )
};

export default AddItems;
