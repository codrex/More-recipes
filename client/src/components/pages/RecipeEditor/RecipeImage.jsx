import React from 'react';
import { connect } from 'react-redux';
import { Image } from 'cloudinary-react';
import PropTypes from 'prop-types';
import toastr from 'toastr';
import { updateImage } from '../../../actions/recipeActions';
import imageParser from '../../../utils/imageParser';
import { DEFAULT_RECIPE_PIX } from '../../../constants';
import toastrConfig from '../../../toastr/toastrConfig';


/**
 * ReactImage uploader component
 */
class RecipeImage extends React.Component {
  /**
   * constructor
   * @return {undefined}
   * @param {object} props
   */
  constructor(props) {
    super(props);
    const { image } = props;
    this.state = {
      image: imageParser(image)
    };

    try {
      this.uploader = this.createUploadWidget();
    } catch (error) {
      // pass
    }
  }

  /**
   * @return {undefined}
   * @param {object} nextProps
   */
  componentWillReceiveProps(nextProps) {
    if (this.props !== nextProps) {
      const { image } = nextProps;
      this.setState({ image: imageParser(image) });
    }
  }

  /**
   * creates cloudinary image upload widget
   * @return {undefined}
   */
  createUploadWidget = () => cloudinary.createUploadWidget(
    {
      cloud_name: process.env.CLOUDINARY_NAME,
      api_key: process.env.CLOUDINARY_API_KEY,
      upload_preset: process.env.CLOUDINARY_PRESET,
      sources: ['local', 'url'],
      client_allowed_formats: ['png', 'gif', 'jpeg'],
      folder: 'recipes',
    },
    (error, result) => {
      const {// eslint-disable-next-line camelcase
        public_id, // eslint-disable-next-line camelcase
        secure_url,
        signature, // eslint-disable-next-line camelcase
        thumbnail_url,
        url
      } = result[0];
      const image = {
        publicId: public_id,
        secureUrl: secure_url,
        signature,
        thumbnailUrl: thumbnail_url,
        url
      };
      this.props.updateImage(JSON.stringify(image));
    }
  )

  /**
   * open cloudinary upload widget
   * @return {undefined}
   */
  openImageUploader = () => {
    try {
      this.uploader.open();
    } catch (error) {
      toastr.error(`Sorry, error encountered, check your
      internet connection and try reloading the page`, 'Error', toastrConfig);
    }
  }

  /**
   * render
   * @return {ReactElement} react component
   */
  render() {
    const { image } = this.state;
    return (
      <div className="image-container">
        <h4 className="items-header-text">
          Recipe image
        </h4>
        <div className="display-4">
          <Image
            cloudName="resycom"
            publicId={image.publicId || DEFAULT_RECIPE_PIX}
            width="400"
            height="500"
            crop="fit"
            className="recipe-card-img"
          />
          <span
            role="button"
            tabIndex="0"
            onClick={this.openImageUploader}
            className="btn bg-secondary"
          >
            {image.url ? 'Change image' : 'Upload image'}
          </span>
        </div>
      </div>
    );
  }
}

RecipeImage.propTypes = {
  image: PropTypes.string.isRequired,
  updateImage: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  image: state.recipe.image
});

export { RecipeImage as PureRecipeImage };
export default connect(mapStateToProps, { updateImage })(RecipeImage);
