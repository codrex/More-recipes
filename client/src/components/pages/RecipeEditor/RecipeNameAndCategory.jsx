import React from 'react';
import { Field, reduxForm, formValueSelector } from 'redux-form';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Form, Select, Input } from '../../common/FormElements';
import { recipeName } from '../../../utils/validator';
import { CATEGORIES } from '../../../constants';
import {
  updateName,
  updateCategory,
  clearValidationError as clearError
} from '../../../actions/recipeActions';

/**
 * RecipeNameAndCategory component
 * @param {object} props
 * @return {React} react element
 */
let RecipeNameAndCategory = (props) => {
  const {
    name,
    recipe,
    validationError,
    clearValidationError
  } = props;
  const {
    name: nameError,
    category: categoryError
  } = validationError;
  return (
    <div>
      <h4 className="items-header-text">
        Recipe name and category
      </h4>
      <Form style={{ zIndex: 4 }}>
        <Field
          component={Input}
          name="name"
          type="text"
          id="recipeName"
          placeholder="Enter recipe name"
          className="add-recipe-input"
          validate={value => recipeName(value, 'recipe name')}
          onBlur={() => props.updateName(name)}
          onFocus={() => clearValidationError({
            name: ''
          })}
          externalError={{
            touched: true,
            error: nameError
          }}
        />
        <Select
          name="category"
          type="text"
          id="category"
          placeholder="Select recipe category"
          className="add-recipe-input"
          options={CATEGORIES}
          value={recipe.category}
          handleChange={({ value }) => {
            props.updateCategory(value);
            clearValidationError({
              category: ''
            });
          }}
          externalError={{
            touched: true,
            error: categoryError
          }}
        />
      </Form>
    </div>
  );
};

RecipeNameAndCategory.defaultProps = {
  recipeCategory: '',
  name: ''
};

RecipeNameAndCategory.propTypes = {
  name: PropTypes.string,
  clearValidationError: PropTypes.func.isRequired,
  recipe: PropTypes.objectOf(PropTypes.shape).isRequired,
  updateName: PropTypes.func.isRequired,
  updateCategory: PropTypes.func.isRequired,
  validationError: PropTypes.objectOf(PropTypes.shape).isRequired,
};

// For testing purpose
const PureRecipeNameAndCategory = RecipeNameAndCategory;
export { PureRecipeNameAndCategory };

RecipeNameAndCategory = reduxForm({
  form: 'recipeNameAndCategory',
})(RecipeNameAndCategory);

const selector = formValueSelector('recipeNameAndCategory');

const mapStateToProps = state => ({
  recipe: state.recipe,
  initialValues: { name: state.recipe.name },
  name: selector(state, 'name'),
  validationError: state.recipeValidationError,
});

export default connect(mapStateToProps, {
  updateCategory,
  updateName,
  clearValidationError: clearError
})(RecipeNameAndCategory);
