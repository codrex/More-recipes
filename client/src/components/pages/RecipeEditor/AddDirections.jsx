import React from 'react';
import { reduxForm } from 'redux-form';
import { connect } from 'react-redux';
import AddItems from './AddItems';
import {
  updateDirections,
  clearValidationError
} from '../../../actions/recipeActions';

/**
 * AddDirections component
 * @param {object} props
 * @return {React} react element
 */
const AddDirections = props => (
  <AddItems
    name="direction"
    directions
    placeholder="Enter direction"
    {...props}
  />
);

const mapStateToProps = state => (
  {
    items: state.recipe.directions,
    externalError: state.recipeValidationError.directions,
    message: state.networkRequest.msg
  }
);

export { AddDirections as PureAddDirections };

export default reduxForm({
  form: 'directionForm'
})(connect(mapStateToProps, {
  sendItemsToStore: updateDirections,
  clearValidationError
})(AddDirections));
