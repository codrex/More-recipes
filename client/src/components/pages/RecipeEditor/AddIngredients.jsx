import React from 'react';
import { reduxForm } from 'redux-form';
import { connect } from 'react-redux';
import AddItems from './AddItems';
import {
  updateIngredients,
  clearValidationError
} from '../../../actions/recipeActions';

const AddIngredients = props => (
  <AddItems
    name="ingredient"
    ingredients
    placeholder="Enter ingredient"
    {...props}
  />
);

const mapStateToProps = state => (
  {
    items: state.recipe.ingredients,
    externalError: state.recipeValidationError.ingredients,
    message: state.networkRequest.msg
  }
);

export default reduxForm({
  form: 'ingredientForm'
})(connect(mapStateToProps, {
  sendItemsToStore: updateIngredients,
  clearValidationError
})(AddIngredients));
