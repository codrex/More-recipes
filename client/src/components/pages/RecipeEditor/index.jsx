import React from 'react';
import PropTypes from 'prop-types';
import Loader from '../../common/Loader';
import Button from '../../common/Button';
import AddDirections from './AddDirections';
import AddIngredients from './AddIngredients';
import RecipeNameAndCategory from './RecipeNameAndCategory';
import RecipeImage from './RecipeImage';
import NotFound from '../../common/NotFound';

/**
 * RecipeEditor
 */
class RecipeEditor extends React.Component {
  /**
   * notFound
   * @return {React} component
   */
  renderNotFound = () => {
    const { hasNotFound } = this.props;
    if (hasNotFound) {
      return <NotFound message="Recipe not found" />;
    }
    return null;
  }

  /**
   * @return {React} react component
   */
  renderBody = () => {
    const {
      loading,
      hasRender,
      submitButtonText,
      handleButtonClick,
      hasNotFound
    } = this.props;
    if (loading || hasNotFound) return null;
    return (
      <div>
        <div
          className="row d-flex justify-content-between align-items-start"
          id="recipe-editor"
          style={{
            margin: '0.5em'
          }}
        >
          <div className="col-xs-12 col-sm-12 col-md-12 col-lg-4 name-category no-padding">
            {hasRender && <RecipeNameAndCategory />}
            <RecipeImage />
          </div>
          <div className="col-xs-12 col-sm-6 col-md-6 col-lg-4 d-flex ingredients no-padding">
            <AddIngredients />
          </div>
          <div className="col-xs-12 col-sm-6 col-md-6 col-lg-4 d-flex directions no-padding">
            <AddDirections />
          </div>
          <Button
            className="btn-secondary btn-lg w-100 text-uppercase add-recipe-btn"
            text={submitButtonText}
            handleClick={handleButtonClick}
          />
        </div>
      </div>
    );
  }

  /**
   * @return {React} react component
   */
  render() {
    const {
      loading,
    } = this.props;
    return (
      <div
        className="container main"
        style={{ maxWidth: 'unset' }}
      >
        {this.renderNotFound()}
        {this.renderBody()}
        {loading && <Loader loading={loading} />}
      </div>
    );
  }
}

RecipeEditor.propTypes = {
  loading: PropTypes.bool.isRequired,
  hasRender: PropTypes.bool.isRequired,
  submitButtonText: PropTypes.string.isRequired,
  handleButtonClick: PropTypes.func.isRequired,
  hasNotFound: PropTypes.bool,
};

RecipeEditor.defaultProps = {
  message: PropTypes.string,
  statusCode: -1,
  hasNotFound: false
};


export default RecipeEditor;
