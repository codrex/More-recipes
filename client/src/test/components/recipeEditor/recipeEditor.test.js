import React from 'react';
import toJson from 'enzyme-to-json';
import RecipeEditor from '../../../components/pages/RecipeEditor';

const props = {
  loading: false,
  hasRender: true,
  submitButtonText: 'post',
  handleButtonClick: jest.fn(),
  hasNotFound: false,
};

describe('Recipe editor component ', () => {
  test('render as expected when component is mounted', () => {
    const component = shallow(
      <RecipeEditor {...props} />
    );
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
  test('expected to match snapshot when recipe was not found', () => {
    const component = shallow(
      <RecipeEditor {...{ ...props, hasNotFound: true }} />
    );
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});
