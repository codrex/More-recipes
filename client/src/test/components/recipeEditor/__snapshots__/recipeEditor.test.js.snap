// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipe editor component  expected to render recipe editor component 1`] = `
<div
  className="container main"
  style={
    Object {
      "maxWidth": "unset",
    }
  }
>
  <div>
    <div
      className="row d-flex justify-content-between align-items-start"
      id="recipe-editor"
      style={
        Object {
          "margin": "0.5em",
        }
      }
    >
      <div
        className="col-xs-12 col-sm-12 col-md-12 col-lg-4 name-category
            no-padding"
      >
        <Connect(ReduxForm) />
        <Connect(RecipeImage) />
      </div>
      <div
        className="col-xs-12 col-sm-6 col-md-6 col-lg-4 d-flex
          ingredients no-padding"
      >
        <ReduxForm />
      </div>
      <div
        className="col-xs-12 col-sm-6 col-md-6 col-lg-4 d-flex
          directions no-padding"
      >
        <ReduxForm />
      </div>
      <Button
        Tag="button"
        className="btn-secondary btn-lg w-100 text-uppercase add-recipe-btn"
        dataTarget=""
        dataToggle=""
        handleClick={[Function]}
        href=""
        id=""
        text="post"
        type="button"
      />
    </div>
  </div>
</div>
`;

exports[`Recipe editor component  should render NotFound component when recipe is not found 1`] = `
<div
  className="container main"
  style={
    Object {
      "maxWidth": "unset",
    }
  }
>
  <NotFound
    message="Recipe not found"
  />
</div>
`;
